src/main/java/org/opensearch/security/privileges/IndexPattern.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+package org.opensearch.security.privileges;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.cluster.metadata.IndexAbstraction;
+import org.opensearch.cluster.metadata.IndexNameExpressionResolver;
+import org.opensearch.security.support.WildcardMatcher;
+
+/**
+ * Aggregates index patterns defined in roles and segments them into patterns using template expressions ("index_${user.name}"),
+ * patterns using date math and plain patterns. This segmentation is needed because only plain patterns can be used
+ * to pre-compute privilege maps. The other types of patterns need to be evaluated "live" during the actual request.
+ */
+public class IndexPattern {
+    private static final Logger log = LogManager.getLogger(IndexPattern.class);
+
+    /**
+     * An IndexPattern which does not match any index.
+     */
+    public static final IndexPattern EMPTY = new IndexPattern(WildcardMatcher.NONE, ImmutableList.of(), ImmutableList.of());
+
+    /**
+     * Plain index patterns without any dynamic expressions like user attributes and date math.
+     * This can be not null. If this instance cannot match any static pattern, this will be WildcardMatcher.NONE.
+     */
+    private final WildcardMatcher staticPattern;
+
+    /**
+     * Index patterns which contain user attributes (like ${user.name})
+     */
+    private final ImmutableList<String> patternTemplates;
+
+    /**
+     * Index patterns which contain date math (like <index_{now}>)
+     */
+    private final ImmutableList<String> dateMathExpressions;
+    private final int hashCode;
+
+    private IndexPattern(WildcardMatcher staticPattern, ImmutableList<String> patternTemplates, ImmutableList<String> dateMathExpressions) {
+        this.staticPattern = staticPattern;
+        this.patternTemplates = patternTemplates;
+        this.dateMathExpressions = dateMathExpressions;
```

Comment:
why require a separate variable for these? Is it for daily-rotating index or something similar?

Reply:
It is a feature of the `index_pattern`  attribute in the roles configuration to support also date math expressions - it has been there from the beginning. 

I guess you could use it for defining that users are just allowed to see the data from today's index, but not from indices in the past. Yet, I find the use of date math for this purpose questionable, as it lacks expressiveness. 

So, this is just for backwards compat.

See also the comment here:

https://github.com/opensearch-project/security/blob/1fef11be731e5e2a3c06c5d565faf23525ffe0d6/src/main/java/org/opensearch/security/privileges/IndexPattern.java#L87-L88

Reply:
TIL. Hmm, I never have used it so didn't understand the comment fully.